### PR TITLE
Add timestamps to position and behavior events

### PR DESCRIPTION
**Pull request by okke-formsma**
When events are queued and later replayed, timing information is currently not preserved. This makes timings of things like nested hold-taps difficult. This PR adds timestamp information to position and behavior events, so timings of nested hold-taps can be handled properly. 

image

previously, this would return hold for the first HT and tap for the other three. Now, all three hold-taps are enabled quickly after one another.

This makes homerow mods much easier to work with. Without this implementation, if you press ctrl+alt+shift TH mods at the same time, you have to wait for 3x the tapping term for all three mods to register. With this PR, you only wait for the tapping_term once!

### CODE REVIEW

**Comment by okke-formsma** [comment]
I found a bug where the timestamp was not sent when an event was received from a peripheral.

File: app/include/drivers/behavior.h [hunk 1]
```diff
@@ -20,7 +20,7 @@
  */
 
 typedef int (*behavior_keymap_binding_callback_t)(struct device *dev, u32_t position, u32_t param1,
-                                                  u32_t param2);
+                                                  u32_t param2, s64_t timestamp);
```
Comment (petejohanson) [inline app/include/drivers/behavior.h:23]: It seems like we're starting to reach the point of "too much" in terms of parameters. I wonder if perhaps instead of these proliferation of parameters, if we need a `struct zmk_key_position_event_data` struct that contains all the data? Then the event definition itself would be just the header and then the data? I start this approach for some of the layer state events I have half completed.

Thoughts?

**Review by petejohanson — COMMENTED** [review]
One API food for thought. Thanks for working on this!

File: app/include/drivers/behavior.h (hunk 1 above)
Comment (innovaker) [reply app/include/drivers/behavior.h:23]: As in ... this?
```
struct zmk_key_position_event_data {
    u32_t position;
    s64_t timestamp;
};
```
or?

Comment (petejohanson) [reply app/include/drivers/behavior.h:23]: Yes, at a minimum. I do wonder if we should figure out a way to "bundle" the two binding parameters too?

Comment (innovaker) [reply app/include/drivers/behavior.h:23]: Parameters as an inner struct or fixed-size array?  What's the cleanest approach to handling one-param and two-param bindings in this context?

Comment (petejohanson) [reply app/include/drivers/behavior.h:23]: I would still expect a fixed two fields for each possible param, that would be `0` and ignored by behaviors that don't use them.

Comment (okke-formsma) [reply app/include/drivers/behavior.h:23]: yes, I had the same thoughts when developing this. Lets pass by value for now to make memory mgmt simple.

Comment (okke-formsma) [reply app/include/drivers/behavior.h:23]:
```
struct zmk_key_position_event_data {
    32_t param1,
    32_t param2,
    u32_t position;
    s64_t timestamp;
};
```

Comment (innovaker) [reply app/include/drivers/behavior.h:23]: Suggest `position` and `timestamp` at the start of the `struct` as those are the least likely to change in the future.

Comment (okke-formsma) [reply app/include/drivers/behavior.h:23]: The overlap with two existing structs is remarkable. Although the first contains a char* instead of a device*, and the second is allocated on the stack and has the wrong lifetime...

```
struct zmk_behavior_binding {
    char *behavior_dev;
    u32_t param1;
    u32_t param2;
};
```
```
struct position_state_changed {
    struct zmk_event_header header;
    u32_t position;
    bool state;
    s64_t timestamp;
};
```

Comment (petejohanson) [reply app/include/drivers/behavior.h:23]: Yeah, so the other issue here is there's the general "event data", and then there's specific "binding info" that should be passed to the behavior we're activating.

I think perhaps we end up with something like:

```
struct zmk_position_state_changed_data {
  u32_t position;
  bool state;
  s64_t timestamp;
}

struct position_state_changed {
  struct zmk_event_header header;
  struct zmk_position_state_changed_data data;
}
```

And then we can pass two things to the behavior, the position state changed data struct, and the binding struct.

Thoughts?

Comment (okke-formsma) [reply app/include/drivers/behavior.h:23]: I'm now implementing the following struct. Not happy with the naming.

```
struct behavior_keymap_binding_data {
    int layer;
    u32_t position;
    s64_t timestamp;
    u32_t param1;
    u32_t param2;
};
```

Comment (petejohanson) [reply app/include/drivers/behavior.h:23]: @okke-formsma Regarding naming `zmk_behavior_binding_event_data` maybe?

Comment (okke-formsma) [reply app/include/drivers/behavior.h:23]: There's also a behavior binding event for sensors which acts slightly different, I'll see if we can use the same data.

Comment (petejohanson) [reply app/include/drivers/behavior.h:23]: I still am not convinced we don't need two structs, both passed to the behavior callbacks, one for the event data itself, separate from the specific behavior being invoked, and the other for the specific binding data that is associated w/ this specific behavior being invoked.

Thoughts?

Comment (okke-formsma) [reply app/include/drivers/behavior.h]: After considering all ideas, I've decided to implement the following:

```
struct zmk_behavior_binding {
    char *behavior_dev;
    u32_t param1;
    u32_t param2;
};

struct zmk_behavior_binding_data {
    int layer;
    u32_t position;
    s64_t timestamp;
};
```

reasoning:
* re-uses existing `struct zmk_behavior_binding`
* "runtime" information stored in `struct zmk_behavior_binding_data`
* not using the position event because this way we can extend the data to include things like the layer information.

File: app/include/drivers/behavior.h [hunk 2]
```diff
@@ -19,8 +20,7 @@
  * (Internal use only.)
  */
 
-typedef int (*behavior_keymap_binding_callback_t)(struct device *dev, u32_t position, u32_t param1,
-                                                  u32_t param2);
+typedef int (*behavior_keymap_binding_callback_t)(struct zmk_behavior_binding *binding, struct zmk_behavior_binding_data data);
 typedef int (*behavior_sensor_keymap_binding_callback_t)(struct device *dev, struct device *sensor,
                                                          u32_t param1, u32_t param2);
```
Comment (petejohanson) [inline app/include/drivers/behavior.h:25]: If we're going to do this, let's refactor this callback to `(struct zmk_behavior_binding *binding, struct device *sensor)` while we're at it.

File: app/include/zmk/behavior.h [hunk 3]
```diff
@@ -10,4 +10,10 @@ struct zmk_behavior_binding {
     char *behavior_dev;
     u32_t param1;
     u32_t param2;
+};
+
+struct zmk_behavior_binding_data {
```
Comment (petejohanson) [inline app/include/zmk/behavior.h:15]: The naming I'm still not 100% happy with.... Thoughts on:

```suggestion
struct zmk_behavior_binding_event_data {
```

File: app/include/zmk/events/keycode-state-changed.h [hunk 4]
```diff
@@ -14,16 +14,20 @@ struct keycode_state_changed {
     u8_t usage_page;
     u32_t keycode;
     bool state;
+    u32_t position;
```
Comment (petejohanson) [inline app/include/zmk/events/keycode-state-changed.h:17]: These events might be synthesized from other place, not just from a specific keybinding, per-se. Why was this needed?

**Review by petejohanson — COMMENTED** [review]
Just one or two minor items, this is really close. Thanks!

File: app/include/zmk/behavior.h (hunk 3 above)
Comment (okke-formsma) [reply app/include/zmk/behavior.h:15]: let's drop the 'data' part?

Comment (okke-formsma) [reply app/include/zmk/behavior.h:15]:
```

struct zmk_behavior_binding_event {
    struct zmk_behavior_binding* binding;
    int layer;
    u32_t position;
    s64_t timestamp;
};
```
this may be an option too...

File: app/include/zmk/events/keycode-state-changed.h (hunk 4 above)
Comment (okke-formsma) [reply app/include/zmk/events/keycode-state-changed.h:17]: we can do without.

**Comment by okke-formsma** [comment]
I've just force-pushed an updated version that includes fixes for all conversations above.

**Review by petejohanson — APPROVED** [review]
Thanks!
